File: timer_manager.js
// timer_manager.js
const blessed = require('blessed');
const fs = require('fs');

class Timer {
    constructor(name, duration) {
        this.name = name;
        this.duration = duration;
        this.elapsed = 0;
        this.running = false;
        this.finished = false;
    }
    remaining() {
        if (this.finished) return 0;
        return Math.max(0, this.duration - this.elapsed);
    }
    update(dt) {
        if (this.running && !this.finished) {
            this.elapsed += dt;
            if (this.elapsed >= this.duration) {
                this.elapsed = this.duration;
                this.running = false;
                this.finished = true;
                return true;
            }
        }
        return false;
    }
}

class TimerManager {
    constructor(saveFile) {
        this.saveFile = saveFile || 'timers.json';
        this.timers = [];
        this.selected = 0;
        this.running = true;
        this.lastUpdate = Date.now();
        this.load();
        if (this.timers.length === 0) {
            this.timers.push(new Timer('Pomodoro', 25*60));
            this.timers.push(new Timer('Break', 5*60));
        }
    }

    addTimer(name, duration) {
        this.timers.push(new Timer(name, duration));
        this.save();
    }

    removeTimer(idx) {
        if (idx < 0 || idx >= this.timers.length) return;
        this.timers.splice(idx, 1);
        if (this.selected >= this.timers.length) this.selected = this.timers.length - 1;
        this.save();
    }

    toggle(idx) {
        if (idx < 0 || idx >= this.timers.length) return;
        const t = this.timers[idx];
        if (t.finished) {
            t.elapsed = 0;
            t.finished = false;
            t.running = true;
        } else {
            t.running = !t.running;
        }
        this.save();
    }

    reset(idx) {
        if (idx < 0 || idx >= this.timers.length) return;
        const t = this.timers[idx];
        t.elapsed = 0;
        t.running = false;
        t.finished = false;
        this.save();
    }

    update() {
        const now = Date.now();
        const dt = (now - this.lastUpdate) / 1000;
        this.lastUpdate = now;
        for (const t of this.timers) {
            if (t.update(dt)) {
                // notification
                screen.displayLine(`!!! Timer '${t.name}' finished !!!`);
            }
        }
    }

    save() {
        const data = this.timers.map(t => ({
            name: t.name,
            duration: t.duration,
            elapsed: t.elapsed,
            running: t.running,
            finished: t.finished
        }));
        fs.writeFileSync(this.saveFile, JSON.stringify(data, null, 2));
    }

    load() {
        try {
            const data = JSON.parse(fs.readFileSync(this.saveFile));
            this.timers = data.map(d => {
                const t = new Timer(d.name, d.duration);
                t.elapsed = d.elapsed;
                t.running = d.running;
                t.finished = d.finished;
                return t;
            });
        } catch (e) {}
    }
}

const screen = blessed.screen({
    smartCSR: true,
    title: 'Timer Manager'
});

const list = blessed.list({
    parent: screen,
    top: 2,
    left: 0,
    width: '100%',
    height: '100%-4',
    keys: true,
    vi: true,
    style: {
        selected: { bg: 'white', fg: 'black' },
        item: { hover: { bg: 'blue' } }
    },
    items: []
});

const status = blessed.text({
    parent: screen,
    bottom: 0,
    left: 0,
    width: '100%',
    height: 3,
    style: { fg: 'white' },
    content: ''
});

const header = blessed.text({
    parent: screen,
    top: 0,
    left: 0,
    width: '100%',
    height: 2,
    style: { fg: 'cyan', bold: true },
    content: '=== Timer Manager ==='
});

const manager = new TimerManager();

function refresh() {
    const items = manager.timers.map((t, i) => {
        const rem = t.remaining();
        const status = t.finished ? ' ✓' : (t.running ? ' ▶' : ' ⏸');
        return `${i+1}. ${t.name} ${status}  ${String(Math.floor(rem/60)).padStart(2,'0')}:${String(Math.floor(rem%60)).padStart(2,'0')} / ${String(Math.floor(t.duration/60)).padStart(2,'0')}:${String(Math.floor(t.duration%60)).padStart(2,'0')}`;
    });
    list.setItems(items);
    list.select(manager.selected);
    status.setContent(`Timers: ${manager.timers.length}   [a]dd  [d]elete  [Space]toggle  [r]eset  [q]uit`);
    screen.render();
}

refresh();

list.on('select', (el, index) => {
    manager.selected = index;
});

screen.key(['q', 'escape'], () => {
    manager.save();
    process.exit(0);
});

screen.key(['a'], () => {
    // simplified: use prompt from blessed
    blessed.prompt({
        parent: screen,
        top: 'center',
        left: 'center',
        width: '50%',
        height: '30%',
        border: { type: 'line' },
        style: { border: { fg: 'cyan' } },
        label: ' Add Timer '
    }, (err, name) => {
        if (name) {
            blessed.prompt({
                parent: screen,
                top: 'center',
                left: 'center',
                width: '50%',
                height: '30%',
                border: { type: 'line' },
                style: { border: { fg: 'cyan' } },
                label: ' Duration (seconds) '
            }, (err, durStr) => {
                if (durStr) {
                    const dur = parseFloat(durStr);
                    if (!isNaN(dur) && dur > 0) {
                        manager.addTimer(name, dur);
                        refresh();
                    }
                }
            });
        }
    });
});

screen.key(['d'], () => {
    if (manager.timers.length > 0) {
        manager.removeTimer(manager.selected);
        refresh();
    }
});

screen.key(['space'], () => {
    if (manager.timers.length > 0) {
        manager.toggle(manager.selected);
        refresh();
    }
});

screen.key(['r'], () => {
    if (manager.timers.length > 0) {
        manager.reset(manager.selected);
        refresh();
    }
});

// Update timer every second
setInterval(() => {
    manager.update();
    refresh();
}, 1000);

list.focus();
screen.render();

File: TimerManager.java
// TimerManager.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.*;

public class TimerManager {
    static class Timer {
        String name;
        double duration;
        double elapsed;
        volatile boolean running;
        volatile boolean finished;
        Timer(String name, double duration) { this.name=name; this.duration=duration; elapsed=0; running=false; finished=false; }
        double remaining() { return finished ? 0 : Math.max(0, duration - elapsed); }
        synchronized boolean update(double dt) {
            if (running && !finished) {
                elapsed += dt;
                if (elapsed >= duration) {
                    elapsed = duration; running=false; finished=true;
                    return true;
                }
            }
            return false;
        }
    }

    private List<Timer> timers = new CopyOnWriteArrayList<>();
    private int selected = 0;
    private boolean running = true;
    private String saveFile = "timers.json";
    private ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
    private long lastUpdate = System.currentTimeMillis();

    public TimerManager() {
        load();
        if (timers.isEmpty()) {
            timers.add(new Timer("Pomodoro", 25*60));
            timers.add(new Timer("Break", 5*60));
        }
    }

    public void addTimer(String name, double duration) {
        timers.add(new Timer(name, duration));
        save();
    }

    public void removeTimer(int idx) {
        if (idx<0 || idx>=timers.size()) return;
        timers.remove(idx);
        if (selected >= timers.size()) selected = timers.size()-1;
        save();
    }

    public void toggle(int idx) {
        if (idx<0 || idx>=timers.size()) return;
        Timer t = timers.get(idx);
        if (t.finished) {
            t.elapsed = 0;
            t.finished = false;
            t.running = true;
        } else {
            t.running = !t.running;
        }
        save();
    }

    public void reset(int idx) {
        if (idx<0 || idx>=timers.size()) return;
        Timer t = timers.get(idx);
        t.elapsed = 0;
        t.running = false;
        t.finished = false;
        save();
    }

    public void update() {
        long now = System.currentTimeMillis();
        double dt = (now - lastUpdate) / 1000.0;
        lastUpdate = now;
        for (Timer t : timers) {
            if (t.update(dt)) {
                System.out.println("\n!!! Timer '" + t.name + "' finished !!!");
            }
        }
    }

    public void save() {
        try (Writer w = new FileWriter(saveFile)) {
            w.write("[");
            for (int i=0; i<timers.size(); i++) {
                Timer t = timers.get(i);
                if (i>0) w.write(",");
                w.write(String.format("{\"name\":\"%s\",\"duration\":%f,\"elapsed\":%f,\"running\":%b,\"finished\":%b}",
                        t.name, t.duration, t.elapsed, t.running, t.finished));
            }
            w.write("]");
        } catch (IOException e) {}
    }

    public void load() {
        try {
            String content = new String(Files.readAllBytes(Paths.get(saveFile)));
            // simple parsing (for demo, use JSON library in production)
            // Здесь упрощённо, в реальном коде используем JSON-библиотеку
        } catch (IOException e) {}
    }

    public void run() throws IOException, InterruptedException {
        System.out.print("\033[?25l"); // hide cursor
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            System.out.print("\033[?25h");
            save();
        }));

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        while (running) {
            // Clear screen
            System.out.print("\033[2J\033[H");
            System.out.println("=== Timer Manager ===");
            System.out.printf("Timers: %d   [a]dd  [d]elete  [Space]toggle  [r]eset  [q]uit\n", timers.size());

            for (int i=0; i<timers.size(); i++) {
                Timer t = timers.get(i);
                double rem = t.remaining();
                String status = t.finished ? " ✓" : (t.running ? " ▶" : " ⏸");
                String line = String.format("%d. %s %s  %02.0f:%02.0f / %02.0f:%02.0f",
                        i+1, t.name, status, rem/60, rem%60, t.duration/60, t.duration%60);
                if (i == selected) System.out.print("\033[7m");
                System.out.println(line);
                System.out.print("\033[0m");
            }
            System.out.println("\nCommands: a=add, d=delete, Space=toggle, r=reset, q=quit");

            // Non-blocking input
            if (System.in.available() > 0) {
                int ch = System.in.read();
                if (ch == 'q' || ch == 27) running = false;
                else if (ch == 'a') {
                    System.out.print("Enter name: ");
                    String name = br.readLine();
                    System.out.print("Enter duration (seconds): ");
                    double dur = Double.parseDouble(br.readLine());
                    addTimer(name, dur);
                } else if (ch == 'd') {
                    if (!timers.isEmpty()) removeTimer(selected);
                } else if (ch == ' ') {
                    if (!timers.isEmpty()) toggle(selected);
                } else if (ch == 'r') {
                    if (!timers.isEmpty()) reset(selected);
                } else if (ch == 27) {
                    // arrow keys: read next chars
                    int c1 = System.in.read();
                    if (c1 == 91) {
                        int c2 = System.in.read();
                        if (c2 == 65 && selected>0) selected--;
                        else if (c2 == 66 && selected<timers.size()-1) selected++;
                    }
                }
            }

            update();
            Thread.sleep(50);
        }
        System.out.print("\033[?25h");
        save();
    }

    public static void main(String[] args) throws Exception {
        TimerManager tm = new TimerManager();
        tm.run();
    }
}
